### DISCUSSION TITLE
Wow!

### DISCUSSION THREAD

**Discussion by nagyrobi** [stas-sl/esphome-sound-level-meter#2] [Ideas]
Awesome work! Please, please submit this component as a PR to ESPHome so it gets a wider audience.

**Comment by stas-sl** [comment]
Thanks :)

I thought about that, however I have some doubts about this... As I mentioned in the comment https://github.com/esphome/feature-requests/issues/1254#issuecomment-1466920227, it is not very composable, i.e. it has one "fat" component that does all. Probably to be included in the main repo, it would require quite significant rework to split it into smaller components that could be chained together, e.g. filters should be extracted as independent smaller components. And implementing this flexible approach managing audio routing, connections, different number of channels, bit depth, buffering, synchronisation, threading, etc. seems like not an easy task. Also, ideally, it needs to be integrated somehow with existing i2s_audio/media_player components, as currently they both work with i2s/audio, but are implemented very differently, different config sections, etc. So I'm considering it more like temporal/intermediate solution.

That said, I'd be happy to continue discussion in mentioned feature-request issue how to make it more unified/flexible/reusable, the solution that could actually be included in the main repo.

**Comment by nagyrobi** [comment]
I don't think that's a problem. Your solution is compact, easy to deploy, out of the box style.

There are many components like this in ESPHome. A lot of sensors and devices speaking _modbus_ protocol have been implemented directly. Only after that, a generic modbus component appeared to modularize the whole system, but the original components are still usable within the system. One can now use them multiple ways.

Same as Time based cover. You can implement it in at least two separate components.

**Reply by stas-sl** [reply]
Well, I can definitely make a PR as is, but I'd like to get at least initial confirmation from core devs (sorry, I'm not sure, what's your role)  who is responsible for such decisions, if there is a chance to be merged, or it should be re-organised slightly/heavily? I don't know what's the best way to communicate with them (besides having discussion on that issue), is it worth pinging someone directly? I've seen recently Jesse mentioned he is also working on i2s/mic support, so I feel like some synchronisation is definitely needed.

**Reply by nagyrobi** [reply]
If you want to get some answers I suggest this:
- make a PR with the code against the DEV branch of ESPHome, and eventually prepare a docs PR too
- enter on Discord on our `dev` channel, post a link to the PR and make a thread there asking for opinions
- if you get positive feedback, finish up the docs PR (crosslink the two PRs so when they get merged they go together)
- if they'd say drop it because a more granular implementation is required, just close the PR

You've got nothing to loose, as you already did the hard work!

**Comment by tom-parkinson** [comment]
Just jumping on here to say a huge thanks for your work on this. It was plug-and-play for me thanks to your clear code and documentation. Great work! I would imagine the ESPHome community would appreciate an accurate, low-cost SPL meter. I certainly do :-) Thanks again for sharing it.
